lib/Connection.php

```diff
@@ -376,15 +382,26 @@ public function tables()
 	 */
 	public function transaction()
 	{
-		if (!$this->connection->beginTransaction())
-			throw new DatabaseException($this);
+        // Transaction started already
+        if (static::$transaction_counter++) {
+            $this->connection->exec('SAVEPOINT trans'.static::$transaction_counter);
+            return;
+        }
+
+        if (!$this->connection->beginTransaction())
+           throw new DatabaseException($this);
 	}
 
 	/**
 	 * Commits the current transaction.
 	 */
 	public function commit()
 	{
+        // Nested transaction simply decrease number
+        if (--static::$transaction_counter) {
```

Comment:
We should release the savepoint, and I think that should be done here.

http://dev.mysql.com/doc/refman/5.7/en/savepoint.html
http://www.postgresql.org/docs/9.1/static/sql-release-savepoint.html